crates/astria-sequencer/src/transaction/unsigned.rs

```diff
@@ -0,0 +1,122 @@
+use astria_proto::sequencer::v1::{
+    unsigned_transaction::Value::AccountsTransaction as ProtoAccountsTransaction,
+    UnsignedTransaction as ProtoUnsignedTransaction,
+};
+use prost::Message as _;
+use serde::{
+    Deserialize,
+    Serialize,
+};
+
+use crate::{
+    accounts::transaction::Transaction as AccountsTransaction,
+    crypto::SigningKey,
+    hash,
+    transaction::signed::Transaction as SignedTransaction,
+};
+
+/// Represents an unsigned sequencer chain transaction.
+/// This type wraps all the different module-specific transactions.
+/// If a new transaction type is added, it should be added to this enum.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[non_exhaustive]
+pub(crate) enum Transaction {
+    AccountsTransaction(AccountsTransaction),
+}
+
+impl Transaction {
+    /// Attempts to encode the unsigned transaction into bytes.
+    #[must_use]
+    pub(crate) fn to_vec(&self) -> Vec<u8> {
+        match &self {
+            Transaction::AccountsTransaction(tx) => ProtoUnsignedTransaction {
+                value: Some(ProtoAccountsTransaction(tx.to_proto())),
+            },
+        }
+        .encode_length_delimited_to_vec()
+    }
+
+    /// Signs the transaction with the given keypair.
+    #[allow(dead_code)]
+    #[must_use]
+    pub(crate) fn sign(self, secret_key: &SigningKey) -> SignedTransaction {
+        let signature = secret_key.sign(&self.hash());
+        SignedTransaction {
+            transaction: self,
+            signature,
+            public_key: secret_key.verification_key(),
+        }
+    }
+
+    pub(crate) fn hash(&self) -> Vec<u8> {
+        hash(&self.to_vec())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use anyhow::{
+        bail,
+        Context as _,
+        Result,
+    };
+    use rand::rngs::OsRng;
+
+    use super::*;
+    use crate::accounts::types::{
+        Address,
+        Balance,
+        Nonce,
+        ADDRESS_LEN,
+    };
+
+    const BOB_ADDRESS: &str = "34fec43c7fcab9aef3b3cf8aba855e41ee69ca3a";
+
+    /// attempts to decode the given hex string into an address.
+    fn address_from_hex_string(s: &str) -> Address {
+        let bytes = hex::decode(s).unwrap();
+        let arr: [u8; ADDRESS_LEN] = bytes.try_into().unwrap();
+        Address::from_array(arr)
+    }
+
+    impl Transaction {
+        /// Attempts to decode an unsigned transaction from the given bytes.
+        ///
+        /// # Errors
+        ///
+        /// - If the bytes cannot be decoded into the prost-generated `UnsignedTransaction` type
+        /// - If the value is missing
+        /// - If the value is not a valid transaction type (ie. does not correspond to any
+        ///   component)
+        fn try_from_slice(bytes: &[u8]) -> Result<Self> {
```

Comment:
As in the other case: I think the proto-deserialization should be made explicit, so this fn would be better as just a `Transaction::try_from_proto`

Reply:
updated, but I kind of like the `try_from_slice` because we won't need to import from `astria_proto` in any crates that eventually use this, as well as `prost::Message` for encoding/decoding. thoughts?